app/services/query-service.js

```diff
@@ -6,13 +6,85 @@ const { createHistoryAwareRetriever } = require('langchain/chains/history_aware_
 const { FakeChatModel } = require('@langchain/core/utils/testing')
 const { AzureAISearchVectorStore } = require('../lib/azure-vector-store')
 const config = require('../config')
+const { logEvent } = require('../insights')
 
 const onFailedAttempt = async (error) => {
   if (error.retriesLeft === 0) {
     throw new Error(`Failed to get embeddings: ${error}`)
   }
 }
 
+const validateResponseLinks = (response) => {
+  let errorMessage = 'validateResponseLinks failed because no response object provided'
+  if (!response) {
+    logEvent(errorMessage)
+
+    return errorMessage
+  }
+
+  const extractLinks = (jsonObj) => {
+    const entries = []
+
+    try {
+      const traverse = (obj, parent) => {
+        if (typeof obj === 'object' && obj !== null) {
+          for (const key in obj) {
+            if (Object.prototype.hasOwnProperty.call(obj, key)) {
+              traverse(obj[key], parent || obj)
+            }
+          }
+        } else if (typeof obj === 'string') {
+          if (obj.includes('http')) {
+            entries.push({ entry: parent, link: obj })
+          }
+        }
+      }
+
+      traverse(jsonObj, null)
+    } catch {}
+
+    const urlRegex = /https?:\/\/[^\s|)]+/g
+
+    const entriesAndLinks = entries.map((entryObj) => {
+      const matches = entryObj.link.match(urlRegex).filter((value, index, self) => self.indexOf(value) === index)
+      return {
+        matches,
+        entry: entryObj.entry
+      }
+    })
+
+    return entriesAndLinks
+  }
+
+  try {
+    // get all links from the response and the context. Use the context as the source of true links.
+    if (!response.answer || !response.context) {
+      errorMessage = 'validateResponseLinks failed because response object does not contain answer or context fields'
+      logEvent(errorMessage)
+
+      return errorMessage
+    }
+
+    const responseEntriesAndLinks = extractLinks(JSON.parse(response.answer))
+    const trueEntriesAndLinks = extractLinks(response.context)
+
+    const invalidLinks = responseEntriesAndLinks.filter((entry) =>
+      entry.matches.some((link) => !trueEntriesAndLinks.some((trueEntry) => trueEntry.matches.includes(link)))
+    )
+
+    if (invalidLinks.length > 0) {
+      errorMessage = 'validateResponseLinks failed because invalid links detected in response objects'
+      logEvent(errorMessage, invalidLinks)
```

Comment:
Logged events are meant to be structured. See `app/lib/events.js`, create a new event `EVENT_HALLUCINATED_LINK`. and use it to log an event, please include the invalid links as a prop